engine/function/src/templates/Numeric.ftl

```diff
@@ -1268,38 +1268,63 @@ public class Numeric {
 
         if (n == 0) {
             return NULL_DOUBLE;
-        } else {
-            // NULL values sorted to beginning of the array.
-            ${pt.primitive}[] copy = sort(values.toArray());
+        }
+
+        ${pt.primitive}[] sorted = values.copyToArray();
+        Arrays.sort(sorted);
 
-            // Determine if there are any NULL in the array.
-            int nullCount = 0;
-            for (int i = 0; i < n; i++) {
-                if (isNull(copy[i])) {
+        int nullStart = -1;
+        int nullCount = 0;
+
+        <#if pt.valueType.isFloat >
+        for (int i = 0; i < n; i++) {
+            final ${pt.primitive} val = sorted[i];
+            if (isNaN(val)) {
+                return Double.NaN; // Any NaN will pollute the result.
+            }
+            if (nullStart == -1) {
+                if (isNull(val)) {
+                    nullStart = i;
                     nullCount++;
-                } else {
-                    break;
                 }
+            } else if (isNull(val)) {
+                nullCount++;
             }
-
-            if (nullCount == 0) {
-                // No NULL, so we can just compute the median and return.
-                if (n % 2 == 0)
-                    return 0.5 * (copy[n / 2 - 1] + copy[n / 2]);
-                else return copy[n / 2];
-            } else if (nullCount < n) {
-                // Some NULL, reduce the count and compute the median of the non-null values.
-                n -= nullCount;
-                if (n % 2 == 0) {
-                    int index = n / 2;
-                    return 0.5 * (copy[n / 2 - 1 + nullCount] + copy[n / 2 + nullCount]);
+        }
+        <#else>
+        for (int i = 0; i < n; i++) {
+            final ${pt.primitive} val = sorted[i];
+            if (nullStart == -1) {
+                if (isNull(val)) {
+                    nullStart = i;
+                    nullCount++;
                 }
-                else return copy[n / 2 + nullCount];
+            } else if (isNull(val)) {
+                nullCount++;
             } else {
-                // All values are NULL.
-                return NULL_DOUBLE;
+                break; // no more NULL possible
             }
         }
+        </#if>
+
+        if (nullCount == n) {
+            return NULL_DOUBLE;
+        } else if (nullCount > 0) {
+            if (nullStart > 0) {
+                // Move the pre-NULL data so we have a contiguous block of non-NULL values.
+                System.arraycopy(sorted, 0, sorted, nullCount, nullStart);
```

Comment:
This data move handles the cases where naive sorting won't clump NULL at the beginning. For example, double arrays containing `Double.NEGATIVE_INFINITY` will sort this value before `NULL_DOUBLE` entries.  This copy moves valid data from before the block of nulls to make a contiguous easily-reasoned block of non-null data.

You could argue that we could carefully manage index computation to correctly avoid the nulls, but this entire case can only be triggered when the vector is of `double` or `float` and contains both `Double.NEGATIVE_INFINITY` and `NULL_DOUBLE` in the same vector.  This logic is easy to follow and trivially correct.

Reply:
Coming from the high-performance numerics world, we would always do the complex indexing to avoid copies like this.

At an absolute minimum, I would comment about the neg inf being at the start.

Reply:
I agree that it's an edge case, but I don't think it's going to be too ugly.